lib/rubocop/cop/style/pattern_argument.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop looks for all? / any? / none? / one? calls where the block
+      # has a single statement and === has been invoked on the receiver so the
+      # === argument can be used as the predicate method argument.
+      #
+      # @example
+      #   # bad
+      #   %w[foo bar foobar].any? { |e| /oo/.match?(e) }
+      #
+      #   [1, 2, 3].any? { |e| (1..10).include?(e) }
+      #
+      #   [1, 2, 3].any? { |e| Set[1, 2, 3] === e }
+      #
+      #   [1, 2, 3].none? do |e|
+      #     Integer === e
+      #   end
+      #
+      #   # good
+      #   %w[foo bar foobar].any?(/oo/)
+      #
+      #   [1, 2, 3].any?(1..10)
+      #
+      #   [1, 2, 3].any?(Set[1, 2, 3])
+      #
+      #   [1, 2, 3].none?(Integer)
+      #
+      class PatternArgument < Cop
+        include RangeHelp
+        extend TargetRubyVersion
+
+        minimum_target_ruby_version 2.5
+
+        def on_block(node)
+          pattern_argument_candidate?(node) do |method, pattern|
+            add_offense(node,
+                        location: method.loc.selector,
+                        message: format(MSG, method: method.method_name, pattern: pattern.source))
+          end
+        end
+
+        def autocorrect(node)
+          lambda do |corrector|
+            _, replacement = pattern_argument_candidate?(node).map(&:source)
+            replacement = "(#{replacement})" unless replacement.match?(/^\(.+\)$/)
+            corrector.replace(block_range_with_space(node), replacement)
+          end
+        end
+
+        private
+
+        MSG = 'Pass `%<pattern>s` as an argument to `%<method>s` instead of a block.'
+
+        def_node_matcher :pattern_argument_candidate?, <<~PATTERN
+          (block $({send csend} _ {:all? :any? :none? :one?}) (:args _) {
```

Comment:
I like it.

In case you don't know, you could create an individual matcher for all possibilites and then have your candidate matcher being "(... {#range? #set? #regexp?})". Not sure if it would be better or not here, just letting you know.

Reply:
I got it working with something like this:

```ruby
def_node_matcher :literal_range?, '(begin ({irange erange} _ _))'
def_node_matcher :new_range?, '(send (const {nil? cbase} :Range) :new _ _)'
def_node_matcher :set?, '(send (const {nil? cbase} :Set) {:[] :new} ...)'
def_node_matcher :literal_regexp?, '(:regexp (:str _) _)'
def_node_matcher :new_regexp?, '(send (const {nil? cbase} :Regexp) :new _)'

def_node_matcher :pattern_argument_candidate?, <<~PATTERN
  (block $({send csend} _ {:all? :any? :none? :one?}) (:args _) {
    (send ${#literal_range? #new_range? #set?} {:=== :include? :member?} ...)
    (send ${#literal_regexp? #new_regexp?} {:=== :=~ :match :match?} ...)
    (send $(const nil? _) :=== _)
    (send lvar :is_a? $(const nil? _))})
PATTERN
```

For some reason I can't combine _literal_ and _new_ node_matchers into a single one, I got "_Couldn't compile due to unbalanced pattern. Pattern: #literal_range? #new_range?_"

Reply:
You probably forgot the `{}`. You must write `def_node_matcher :range?, '{#literal_range? #new_range?}'`

Reply:
Thanks! Just split them, I think that way it's easier to read.